### ISSUE TITLE
split(), Tuple and GPU not working nicely together

### ISSUE THREAD

**Issue by kgnk** [halide/Halide#2056]
The code below causes the extent of thread_x to be the full width of the input. Asserting the width is modulo the vector width makes the extent reasonable again.

```
#include "Halide.h"

using namespace Halide;
Var x("x"), y("y"), bx("bx"), by("by"), tx("tx"), ty("ty");

void generate(char const *tag, Expr wd, Expr ht, Target const &target)
{
    ImageParam input(Float(32), 2, std::string("input") + tag);

    input
        .dim(0).set_bounds(0, wd)
        .dim(1).set_bounds(0, ht).set_stride(wd)
        ;

    Func tuple(std::string("tuple") + tag);
    tuple(x, y) = Tuple(input(x, y), input(x, y));

    Func consumer(std::string("consumer") + tag);
    consumer(x, y) = input(x, y) + tuple(x, y)[0];

    // Schedule

    consumer
        .bound( x, 0, wd )
        .bound( y, 0, ht )
        .vectorize( x, 4, TailStrategy::ShiftInwards )
        .tile( x, y, bx, by, tx, ty, 16, 16, TailStrategy::ShiftInwards )
        .gpu_blocks( bx, by )
        .gpu_threads( tx, ty )
        .compute_root()
        ;

    tuple
        .vectorize( x, 4, TailStrategy::RoundUp )
        .compute_at( consumer, bx )
        .gpu_threads( x, y )
        ;

    consumer.compile_to_lowered_stmt
        ( std::string("consumer") + tag + ".stmt"
        , consumer.infer_arguments()
        , Text
        , target
        );
 }

int main()
{
    Param<int> wdVar("wd"), htVar("ht");
    int wdFix = 1022, htFix = 1022;

    Target target( "host-opencl-no_asserts-no_bounds_query" );
    generate( "Var", wdVar, htVar, target );
    generate( "Fix", wdFix, htFix, target );
    generate( "Mod", wdVar / 4 * 4, htVar, target );

    return 0;
}
```

**Comment by kgnk** [comment]
I've tried a few things to isolate the issue. I've updated the title appropriately. It seems to be the combination of two splits (making three levels: vector, thread, block), Tuple computed at the block level, a width that may not be modulo the split size and gpu_blocks/gpu_threads. If I change any of these, the loops look good.

**Comment by psuriana** [comment]
That looks like a bug. I'll take a look.

**Comment by psuriana** [comment]
Fixed in #2061
